### ISSUE TITLE
temp input plugin: sensor names not unique on a multi-socket machine

### ISSUE THREAD

**Issue by gsauthof** [influxdata/telegraf#7222]
### Relevant telegraf.conf:

```toml
[[inputs.temp]]
```

### System info:

- 1.13.4
- CentOS 7
- machine with multiple CPU sockets

### Steps to reproduce:

1. enable temp plugin
2. run telegraf on a machine with multiple sockets in `--test` mode

### Expected behavior:

`temp,host=example.org,sensor=coretemp*_core*_input temp=`

line prefixes are unique

### Actual behavior:

There are duplicated prefixes, e.g.:

```
[..]
temp,host=example.org,sensor=coretemp_core0_input temp=[..]
[..]
temp,host=example.org,sensor=coretemp_core0_input temp=[..]
[..]
```

### Additional info:

This is caused by the read code in

https://github.com/shirou/gopsutil/blob/3964e5b57942b45396df794baa03df27a667f895/host/host_linux.go#L447-L510

that generates non-unique keys because on a multi-socket system each socket gets a separate directory under `/sys/class/hwmon`  where the name and labels inside don't differ.

Thus, as is, only the core temperatures of the last socket are actually stored in an influxdb.

**Comment by danielnelson** [comment]
I opened an issue with gopsutil, if we received a unique identifier it would certainly make the fix easier than attempting to do it in Telegraf.

https://github.com/shirou/gopsutil/issues/854

**Comment by powersj** [comment]
This was fixed in https://github.com/shirou/gopsutil/pull/485.

**Comment by gsauthof** [comment]
@powersj wrote:

> This was fixed in [shirou/gopsutil#485](https://github.com/shirou/gopsutil/pull/485).

No, it wasn't. Note that the issue @danielnelson opened upstream is still open.

On a dual socket (2 x 32 cores, no hyperthreading) server I'm still seeing duplicated core-temperature sensor names, using the latest telegraf release:

```
$ ./telegraf-1.37.2/usr/bin/telegraf --config <(echo '[[inputs.temp]]') --test | grep core \
      | sed 's/host=[^,]\+,/host=example.org,/' | sed 's/temp=.\+$//'  | sort | uniq -c
      2 > temp,host=example.org,sensor=coretemp_core_0
      2 > temp,host=example.org,sensor=coretemp_core_1
      2 > temp,host=example.org,sensor=coretemp_core_10
      2 > temp,host=example.org,sensor=coretemp_core_11
      2 > temp,host=example.org,sensor=coretemp_core_12
      2 > temp,host=example.org,sensor=coretemp_core_13
      2 > temp,host=example.org,sensor=coretemp_core_14
      2 > temp,host=example.org,sensor=coretemp_core_15
      2 > temp,host=example.org,sensor=coretemp_core_16
      2 > temp,host=example.org,sensor=coretemp_core_17
      2 > temp,host=example.org,sensor=coretemp_core_18
      2 > temp,host=example.org,sensor=coretemp_core_19
      2 > temp,host=example.org,sensor=coretemp_core_2
      2 > temp,host=example.org,sensor=coretemp_core_20
      2 > temp,host=example.org,sensor=coretemp_core_21
      2 > temp,host=example.org,sensor=coretemp_core_22
      2 > temp,host=example.org,sensor=coretemp_core_23
      2 > temp,host=example.org,sensor=coretemp_core_24
      2 > temp,host=example.org,sensor=coretemp_core_25
      2 > temp,host=example.org,sensor=coretemp_core_26
      2 > temp,host=example.org,sensor=coretemp_core_27
      2 > temp,host=example.org,sensor=coretemp_core_28
      2 > temp,host=example.org,sensor=coretemp_core_29
      2 > temp,host=example.org,sensor=coretemp_core_3
      2 > temp,host=example.org,sensor=coretemp_core_30
      2 > temp,host=example.org,sensor=coretemp_core_31
      2 > temp,host=example.org,sensor=coretemp_core_4
      2 > temp,host=example.org,sensor=coretemp_core_5
      2 > temp,host=example.org,sensor=coretemp_core_6
      2 > temp,host=example.org,sensor=coretemp_core_7
      2 > temp,host=example.org,sensor=coretemp_core_8
      2 > temp,host=example.org,sensor=coretemp_core_9
      1 > temp,host=example.org,sensor=coretemp_package_id_0
      1 > temp,host=example.org,sensor=coretemp_package_id_1
```

---

Some details about the sensor topology:

The two hwmon directories can be differentiated by looking at their links:

```
$ cd /sys/class/hwmon
ls -l | grep core
lrwxrwxrwx 1 root root 0 Feb 12 09:50 hwmon11 -> ../../devices/platform/coretemp.0/hwmon/hwmon11
lrwxrwxrwx 1 root root 0 Feb 12 09:50 hwmon12 -> ../../devices/platform/coretemp.1/hwmon/hwmon12
```

Looking at the labels is not sufficient:

```
$ grep . hwmon1[12] -r --include '*label*' | sort -V
hwmon11/temp1_label:Package id 0
hwmon11/temp2_label:Core 0
hwmon11/temp3_label:Core 1
hwmon11/temp4_label:Core 2
hwmon11/temp5_label:Core 3
[..]
hwmon11/temp9_label:Core 7
hwmon11/temp10_label:Core 8
hwmon11/temp11_label:Core 9
[..]
hwmon11/temp29_label:Core 27
hwmon11/temp30_label:Core 28
hwmon11/temp31_label:Core 29
hwmon11/temp32_label:Core 30
hwmon11/temp33_label:Core 31
hwmon12/temp1_label:Package id 1
hwmon12/temp2_label:Core 0
hwmon12/temp3_label:Core 1
hwmon12/temp4_label:Core 2
[..]
hwmon12/temp29_label:Core 27
hwmon12/temp30_label:Core 28
hwmon12/temp31_label:Core 29
hwmon12/temp32_label:Core 30
hwmon12/temp33_label:Core 31
```

Each coretemp sensor has the following files available:


```
$ ls hwmon12 | sort -V | head -30
device
name
power
subsystem
temp1_crit
temp1_crit_alarm
temp1_input
temp1_label
temp1_max
temp2_crit
temp2_crit_alarm
temp2_input
temp2_label
temp2_max
temp3_crit
temp3_crit_alarm
temp3_input
temp3_label
temp3_max
```

It's also suboptimal that telegraf reads many files whose content it doesn't actually use:

```
strace -y -f -o t.log  ./telegraf-1.37.2/usr/bin/telegraf --config <(echo '[[inputs.temp]]') --test
grep 'open.*hwmon1[12]' t.log
354513 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11>
354513 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon12", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.1/hwmon/hwmon12>
354513 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/name", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/name>
354513 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/temp10_label", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/temp10_label>
354513 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/temp10_input", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/temp10_input>
354513 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11>
354513 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/temp10_crit", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/temp10_crit>
354513 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/temp10_crit_alarm", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/temp10_crit_alarm>
354513 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/temp10_input", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/temp10_input>
354513 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/temp10_label", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/temp10_label>
354499 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/temp10_max", O_RDONLY|O_CLOEXEC <unfinished ...>
354499 <... openat resumed>)            = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/temp10_max>
354499 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/name", O_RDONLY|O_CLOEXEC <unfinished ...>
354499 <... openat resumed>)            = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/name>
354499 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/temp11_label", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/temp11_label>
354499 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/temp11_input", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/temp11_input>
354499 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11", O_RDONLY|O_CLOEXEC <unfinished ...>
354499 <... openat resumed>)            = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11>
354499 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/temp11_crit", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/temp11_crit>
354499 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/temp11_crit_alarm", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/temp11_crit_alarm>
354499 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/temp11_input", O_RDONLY|O_CLOEXEC <unfinished ...>
354499 <... openat resumed>)            = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/temp11_input>
354499 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/temp11_label", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/temp11_label>
354499 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/temp11_max", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/temp11_max>
354499 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/name", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/name>
354499 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/temp12_label", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/temp12_label>
354499 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/temp12_input", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/temp12_input>
354499 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11>
354499 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/temp12_crit", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/temp12_crit>
354499 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/temp12_crit_alarm", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/temp12_crit_alarm>
354499 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/temp12_input", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/temp12_input>
354499 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/temp12_label", O_RDONLY|O_CLOEXEC <unfinished ...>
354499 <... openat resumed>)            = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/temp12_label>
354499 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/temp12_max", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/temp12_max>
354499 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/name", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/name>
354499 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/temp13_label", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/temp13_label>
354499 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/temp13_input", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/temp13_input>
354499 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11>
354499 openat(AT_FDCWD</local/wu224>, "/sys/class/hwmon/hwmon11/temp13_crit", O_RDONLY|O_CLOEXEC) = 7</sys/devices/platform/coretemp.0/hwmon/hwmon11/temp13_crit>
```

Meaning it reads all the `*_crit` and `*_max` files although it doesn't report those values.

This is bad because every read requires the kernel to schedule a `rdmsr` instruction via an interrupt on the affected core.

Furthermore, it's weird that files related to the 2nd socket are read twice per collection event:

```
$ grep 'open.*hwmon1[12]' t.log | grep -o '".*)' | sort | uniq -c | sort -k1,1 -n
      1 "/sys/class/hwmon/hwmon11/temp10_crit_alarm", O_RDONLY|O_CLOEXEC)
      1 "/sys/class/hwmon/hwmon11/temp10_crit", O_RDONLY|O_CLOEXEC)
      1 "/sys/class/hwmon/hwmon11/temp11_crit_alarm", O_RDONLY|O_CLOEXEC)
      1 "/sys/class/hwmon/hwmon11/temp11_crit", O_RDONLY|O_CLOEXEC)
      1 "/sys/class/hwmon/hwmon11/temp11_input", O_RDONLY|O_CLOEXEC)
      1 "/sys/class/hwmon/hwmon11/temp11_max", O_RDONLY|O_CLOEXEC)
      1 "/sys/class/hwmon/hwmon11/temp12_crit_alarm", O_RDONLY|O_CLOEXEC)
      1 "/sys/class/hwmon/hwmon11/temp12_crit", O_RDONLY|O_CLOEXEC)
      1 "/sys/class/hwmon/hwmon11/temp12_label", O_RDONLY|O_CLOEXEC)
      1 "/sys/class/hwmon/hwmon11/temp12_max", O_RDONLY|O_CLOEXEC)
      1 "/sys/class/hwmon/hwmon11/temp13_crit_alarm", O_RDONLY|O_CLOEXEC)
      1 "/sys/class/hwmon/hwmon11/temp13_crit", O_RDONLY|O_CLOEXEC)
      1 "/sys/class/hwmon/hwmon11/temp13_input", O_RDONLY|O_CLOEXEC)
      1 "/sys/class/hwmon/hwmon11/temp13_max", O_RDONLY|O_CLOEXEC)
      1 "/sys/class/hwmon/hwmon11/temp14_crit_alarm", O_RDONLY|O_CLOEXEC)
      1 "/sys/class/hwmon/hwmon11/temp14_crit", O_RDONLY|O_CLOEXEC)
      1 "/sys/class/hwmon/hwmon11/temp14_max", O_RDONLY|O_CLOEXEC)
      1 "/sys/class/hwmon/hwmon11/temp15_crit_alarm", O_RDONLY|O_CLOEXEC)
      1 "/sys/class/hwmon/hwmon11/temp15_crit", O_RDONLY|O_CLOEXEC)
      1 "/sys/class/hwmon/hwmon11/temp15_max", O_RDONLY|O_CLOEXEC)
[..]
      2 "/sys/class/hwmon/hwmon12/temp2_label", O_RDONLY|O_CLOEXEC)
      2 "/sys/class/hwmon/hwmon12/temp30_label", O_RDONLY|O_CLOEXEC)
      2 "/sys/class/hwmon/hwmon12/temp31_input", O_RDONLY|O_CLOEXEC)
      2 "/sys/class/hwmon/hwmon12/temp31_label", O_RDONLY|O_CLOEXEC)
      2 "/sys/class/hwmon/hwmon12/temp32_label", O_RDONLY|O_CLOEXEC)
      2 "/sys/class/hwmon/hwmon12/temp33_input", O_RDONLY|O_CLOEXEC)
      2 "/sys/class/hwmon/hwmon12/temp33_label", O_RDONLY|O_CLOEXEC)
      2 "/sys/class/hwmon/hwmon12/temp3_input", O_RDONLY|O_CLOEXEC)
      2 "/sys/class/hwmon/hwmon12/temp3_label", O_RDONLY|O_CLOEXEC)
      2 "/sys/class/hwmon/hwmon12/temp4_input", O_RDONLY|O_CLOEXEC)
      2 "/sys/class/hwmon/hwmon12/temp4_label", O_RDONLY|O_CLOEXEC)
      2 "/sys/class/hwmon/hwmon12/temp5_input", O_RDONLY|O_CLOEXEC)
      2 "/sys/class/hwmon/hwmon12/temp5_label", O_RDONLY|O_CLOEXEC)
      2 "/sys/class/hwmon/hwmon12/temp6_input", O_RDONLY|O_CLOEXEC)
      2 "/sys/class/hwmon/hwmon12/temp6_label", O_RDONLY|O_CLOEXEC)
      2 "/sys/class/hwmon/hwmon12/temp7_input", O_RDONLY|O_CLOEXEC)
      2 "/sys/class/hwmon/hwmon12/temp7_label", O_RDONLY|O_CLOEXEC)
      2 "/sys/class/hwmon/hwmon12/temp8_input", O_RDONLY|O_CLOEXEC)
      2 "/sys/class/hwmon/hwmon12/temp8_label", O_RDONLY|O_CLOEXEC)
      2 "/sys/class/hwmon/hwmon12/temp9_input", O_RDONLY|O_CLOEXEC)
      2 "/sys/class/hwmon/hwmon12/temp9_label", O_RDONLY|O_CLOEXEC)
     31 "/sys/class/hwmon/hwmon11", O_RDONLY|O_CLOEXEC)
     31 "/sys/class/hwmon/hwmon12/name", O_RDONLY|O_CLOEXEC)
     32 "/sys/class/hwmon/hwmon11/name", O_RDONLY|O_CLOEXEC)
     33 "/sys/class/hwmon/hwmon12", O_RDONLY|O_CLOEXEC)
```

Also weird that directories/files `/sys/class/hwmon/hwmon1[12]` and `/sys/class/hwmon/hwmon1[12]/name` are opened over 30 times.

Server details:

- OS: RHEL 9.6
-  CPU: dual socket, 2 x Intel Xeon Platinum 8462Y+ (Hyperthreading disabled)

---

Since the reading of some sensors has side-effects (such adding latency to a device/core) it would be great if the plugin would allow to exclude some sensors. For example, to exclude all core-temperatures and only read the temperatures of the physical CPU packages.

Also, even if the temperature plugin would report crit/max temperatures at some point, it should be optional to save space.
And one could also argue about how much sense it makes to read out those periodically, since those values are constants, usually.
